src/MySqlDatabase.php

```diff
@@ -0,0 +1,274 @@
+    }
+
+    /**
+     * Sets the name of the table where migration versions are logged.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable $logTable The name of the version table.
+     */
+    protected function _setLogTableName($logTable)
+    {
+        $this->logTable = $this->_normalizeString($logTable);
+    }
+
+    /**
+     * Retrieves the name of the version column in the log table.
+     *
+     * @since [*next-version*]
+     *
+     * @return string The name of the version column in the log table.
+     */
+    protected function _getLogTableVersionColumn()
+    {
+        return $this->versionColumn;
+    }
+
+    /**
+     * Sets the name of the version column in the log table.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable $versionColumn The name of the version column in the log table.
+     */
+    protected function _setLogTableVersionColumn($versionColumn)
+    {
+        $this->versionColumn = $this->_normalizeString($versionColumn);
+    }
+
+    /**
+     * Retrieves the name of the version column in the log table.
+     *
+     * @since [*next-version*]
+     *
+     * @return string The name of the version column in the log table.
+     */
+    protected function _getLogTableStatusColumn()
+    {
+        return $this->statusColumn;
+    }
+
+    /**
+     * Sets the name of the version column in the log table.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable $statusColumn The name of the version column in the log table.
+     */
+    protected function _setLogTableStatusColumn($statusColumn)
+    {
+        $this->statusColumn = $this->_normalizeString($statusColumn);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function executeSql($sql)
+    {
+        $this->getDbDriver()->execute($sql);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function createDatabase()
+    {
+        $this->executeSql(
+            $this->_formatSql('CREATE SCHEMA IF NOT EXISTS `{db}` DEFAULT CHARACTER SET utf8 ;')
+        );
+        $this->executeSql(
+            $this->_formatSql('USE `{db}`')
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function dropDatabase()
+    {
+        $this->executeSql(
+            $this->_formatSql('DROP DATABASE `{db}`')
+        );
+    }
+
+    /**
+     * Retrieves the name of the database in a given URI.
+     *
+     * @since [*next-version*]
+     *
+     * @param Uri $uri The URI.
+     *
+     * @return string The extracted name of the database.
+     */
+    public static function getDatabaseNameFromUri(Uri $uri)
+    {
+        return preg_replace('~^/~', '', $uri->getPath());
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public static function prepareEnvironment(Uri $uri)
```

Comment:
Not going to fix this to accept an interface?

Reply:
AFAIK, an overridden method signature can't loosen a parameter type.

Reply:
In that case, very sad.

Reply:
What I _can_ change is the param type hint for `getDatabaseNameFromUri()`.

Reply:
Looks like this is now done in fac25980ad29dd7a9735d81b88df6e835ad6bec9.

Reply:
Another question, @mecha: why is there need to "prepare environment", in the sense that involves using globals (e.g. a static factory method)?

Reply:
/shrug